addons/l10n_ro_edi_stock/models/l10n_ro_edi_stock_document.py

```diff
@@ -0,0 +1,100 @@
+import requests
+import re
+
+from odoo import api, fields, models, _
+
+DOCUMENT_STATES = [
+    ('stock_sent', "Sent"),
+    ('stock_sending_failed', "Error"),
+    ('stock_validated', 'Validated'),
+]
+
+SCHEMATRON_ERROR_ID_PATTERN = r'BR-(?:CL-)?\d{3}'
+
+ETRANSPORT_URLS = {
+    'test': 'https://api.anaf.ro/test/ETRANSPORT/ws/v1',
+    'prod': 'https://api.anaf.ro/prod/ETRANSPORT/ws/v1'
+}
+
+
+def _cleanup_errors(errors: list[str]):
+    def _cleanup_schematron_error(error: str) -> str:
+        for part in error.split('; '):
+            key, value = part.split('=', maxsplit=1)
+            if key == 'textEroare':
+                return value.strip()
+
+    return [_cleanup_schematron_error(err) if re.search(SCHEMATRON_ERROR_ID_PATTERN, err) else err.strip() for err in errors]
+
+
+def _make_etransport_request(company, endpoint: str, method: str, session, data=None):
+    api_env = 'test' if company.l10n_ro_edi_test_env else 'prod'
+    url = f"{ETRANSPORT_URLS[api_env]}/{endpoint}"
+    headers = {
+        'Content-Type': 'application/xml',
+        'Authorization': f'Bearer {company.l10n_ro_edi_access_token}',
+    }
+
+    # encode data to utf-8 because it could contain some Romanian characters that are not part of latin-1
+    if data:
+        data = data.encode()
+
+    try:
+        response = session.request(method=method, url=url, data=data, headers=headers, timeout=10)
+    except requests.HTTPError as e:
+        return {'error': str(e)}
+
+    match response.status_code:
+        case 404:
+            return {'error': response.json()['message']}
+        case 403:
+            return {'error': _("Access token is forbidden.")}
+        case 204:
+            return {'error': _("You reached the limit of requests. Please try again later.")}
+
+    try:
+        response.raise_for_status()
+    except requests.HTTPError as e:
```

Comment:
`raise_for_status` could also return a **Timeout** error or a **TooManyRedirection** error. If the goal is to catchem'all, use `requests.exceptions.RequestException` which is the top-level request exception implementation :+1: